spinalcordtoolbox/reports/qc2.py

```diff
@@ -204,263 +205,300 @@ def rescale_grid(
+            },
+            shape=shape,
+            axis_labels=(("A", "P"), ("L", "R")),
+        )
 
-    # Save the figure as a png
-    fig.savefig(str(path), format='png', transparent=True)
+    def center_patches(
+        self, seg: Image, shape: tuple[int, int] = (30, 30)
+    ) -> 'SlicingSpec':
+        """
+        A re-centered and cropped slicing spec, meant for axial views.
+
+        Each slice is individually centered around the center of mass of the
+        given segmentation before being cropped (in both axes).
+        """
+        # Recenter each slice around its center of mass, and interpolate missing
+        # slices. We use a single (N, 3) array so that we can inf_nan_fill.
+        offsets = np.array(list(self.offsets.values()))
+        slices_seg = self.get_slices(seg, order=1)
+        for offset, slice_seg in zip(offsets, slices_seg.values()):
+            offset[1:3] += ndimage.center_of_mass(slice_seg)
+        inf_nan_fill(offsets[:, 1])
+        inf_nan_fill(offsets[:, 2])
+
+        # Take into account the size of the cropping rectangle.
+        offsets[:, 1:3] -= (np.array(shape) - 1) / 2
+
+        return SlicingSpec(
+            affine=self.affine,
+            offsets=dict(zip(self.offsets.keys(), offsets)),
+            shape=shape,  # new shape
+            axis_labels=self.axis_labels,
+        )
+
+    @staticmethod
+    def full_sagittal(img: Image, p_resample: float = 0.6) -> 'SlicingSpec':
+        """
+        A slicing spec to extract full sagittal slices.
+
+        The field of view is given by `img`.
+        The number of slices is the same as `img` in the R-L axis,
+        but the other two axes are resampled to `p_resample` mm.
+        The slice labels correspond to the original orientation of `img`,
+        but the output slices are always ordered and oriented in "RSP".
+        """
+        # Standardize to RSP with a fixed resolution for each SP slice.
+        shape, affine = img.data.shape, img.affine
+        shape, affine = reorient_grid(shape, affine, "RSP")
+        shape, affine = rescale_grid(shape, affine, [None, p_resample, p_resample])
+
+        # Split up the 3D image shape into a number of slices and a 2D slice shape.
+        num_slices, shape = shape[0], shape[1:3]
+
+        # The slice labels in the original orientation of the input image.
+        slice_labels = [str(i) for i in range(num_slices)]
+        if "L" in img.orientation:
+            slice_labels.reverse()
+
+        return SlicingSpec(
+            affine=affine,
+            offsets={
+                label: np.array([i, 0, 0], dtype=np.float64)
+                for i, label in enumerate(slice_labels)
+            },
+            shape=shape,
+            axis_labels=(("S", "I"), ("P", "A")),
+        )
+
+    def center_columns(self, seg: Image) -> 'SlicingSpec':
+        """
+        A re-centered and cropped slicing spec, meant for sagittal views.
```

Comment:
It looks like the `center_` methods also seem to have a requirement that a `full_` method needs to have been run first, so maybe we could add validation for that too?